### PR TITLE
[bin] onnx_streaming and onnx_non_streaming

### PR DESCRIPTION
**Pull request by xingchensong**
I find it quite easy to export onnx (either streaming model or non-streaming model) within our recently developed new cache API. The most tricky thing for onnx export is to ensure that all if-else branches in `forward_chunk` will always take the same path with a given decoding setup (i.e., 16/-1, 16/4, 16/0 or -1/-1) (this will turn the dynamic if-else branch into a static branch and `next_cache_start` will be a fixed value). To do this, we can simply adjust the shape of cache & mask tensors for the **first** input chunk according to `chunk_size` and `num_left_chunks`.

# TODO
- [x] check consistency between onnx_model and pytorch_model on random data
- [ ] check CER on aishell-1 testset

# Q&A
1. Why don't we need `jit.script` for if-else branches?
As mentioned above, we force our model to select the same branch for every chunk.

2. Why don't we need `jit.script` for tensor slice?
**From @Mddct ** : ONNX supports dynamic slice for opset>=13, see https://github.com/onnx/onnx/blob/main/docs/Operators.md#Slice

```py
# Author: @Mddct 
import torch
from torch import nn

class BadFirst(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x: torch.Tensor, index: torch.Tensor):

        x_firsts = x[:index]
        return x_firsts


m = BadFirst().eval()
x = torch.ones(44)
index = torch.tensor(0, dtype=torch.int)

res = m(x, index=index) # this works
torch.onnx.export(m, (x, index),  "m.onnx",  opset_version=14)
```

### CODE REVIEW

**Comment by robin1001** [comment]
@xingchensong , please cowork with @lzhin on this PR.

**Comment by robin1001** [comment]
@lzhin, please review and verify the PR.

**Comment by lzhin** [comment]
@robin1001, ok, I'll do it soon.

**Comment by Mddct** [comment]
May we can tun if-else branch code  to tensor operation for export onnx
```python
zeor_tensor = torch.tensor(0)

lt_zero = torch.less(required_cache_size, zeor_tensor)
gt_zero = torch.greater(required_cache_size, zeor_tensor)

tmp_cur_offset = attention_key_size - required_cache_size
tmp_cur_offset = torch.where(torch.greater(tmp_cur_offset, 0), 
                             tmp_cur_offset, 
                             0)

next_cache_size = torch.where(lt_zero, 
                              torch.tensor(0),
                              attention_key_size
                              )

next_cache_size = torch.where(gt_zero,
                              tmp_cur_offset,
                              next_cache_size,
                              )
```

File: wenet/bin/export_onnx_cpu.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+    inputs = (chunk, offset, required_cache_size,
+              att_cache, cnn_cache, att_mask)
+    print("\t\t\033[32mchunk.size(): {}\033[0m\n".format(chunk.size()),
+          "\t\t\033[32moffset: {}\033[0m\n".format(offset),
+          "\t\t\033[32mrequired_cache: {}\033[0m\n".format(required_cache_size),
+          "\t\t\033[32matt_cache.size(): {}\033[0m\n".format(att_cache.size()),
+          "\t\t\033[32mcnn_cache.size(): {}\033[0m\n".format(cnn_cache.size()),
+          "\t\t\033[32matt_mask.size(): {}\033[0m\n".format(att_mask.size()))
+
+    print("\t\033[32mStage-1.2: torch.onnx.export\033[0m")
```
Comment (robin1001) [inline wenet/bin/export_onnx_cpu.py:134]: replace print("\t\033[32m{}\033[0m") with a function

Comment (robin1001) [reply wenet/bin/export_onnx_cpu.py:134]: Can we split the whole processing into functions, seems it is too long now.
I think we can move the verfication code into a new function.

**Comment by robin1001** [comment]
I got another question, how can we save `eos`, `sos`, `subsampling_rage`, `right context`, etc information? what's the best way to do that?

**Comment by xingchensong** [comment]
> I got another question, how can we save `eos`, `sos`, `subsampling_rage`, `right context`, etc information? what's the best way to do that?

from train.yaml and initialized model

https://github.com/wenet-e2e/wenet/pull/1023/files#diff-e4274fd35d26689b022f06cb30cd14be040f8075c97a36521684c02dad9412ecR83-R91

https://github.com/wenet-e2e/wenet/pull/1023/files#diff-e4274fd35d26689b022f06cb30cd14be040f8075c97a36521684c02dad9412ecR278

File: wenet/bin/export_onnx_cpu.py (hunk 1 above)
Comment (xingchensong) [reply wenet/bin/export_onnx_cpu.py]: done

**Comment by Mddct** [comment]
> I got another question, how can we save `eos`, `sos`, `subsampling_rage`, `right context`, etc information? what's the best way to do that?

can we do this：
torch onnx export four onnx models：
1 encoder
2 decoder
3 ctc decoder
4 meta （just some constant）

and then concat those four onnx graps to one graph：https://github.com/scailable/sclblonnx/blob/master/sclblonnx/merge.py#L213

we can call session.run on a graph using different input names

**Comment by Mddct** [comment]
> > I got another question, how can we save `eos`, `sos`, `subsampling_rage`, `right context`, etc information? what's the best way to do that?
> 
> can we do this： torch onnx export four onnx models： 1 encoder 2 decoder 3 ctc decoder 4 meta （just some constant）
> 
> and then concat those four onnx graps to one graph：https://github.com/scailable/sclblonnx/blob/master/sclblonnx/merge.py#L213
> 
> we can call session.run on a graph using different input names

we can't run session using different input namse, because session is binding to a model with fixed input names.

But i find a keyword in onnx proto: 
Optional Metadata：https://github.com/onnx/onnx/blob/main/docs/IR.md
Maybe we can keep sos ... here

**Comment by xingchensong** [comment]
we can write and read metadatas like this:

```py
# write
onnx_encoder = onnx.load(encoder_outpath)
for (k, v) in args.items():
    meta = onnx_encoder.metadata_props.add()
    meta.key, meta.value = str(k), str(v)
onnx.save(onnx_encoder, encoder_outpath)

# read
ort_session = onnxruntime.InferenceSession(encoder_outpath)
meta = ort_session.get_modelmeta()
print("\t\tcustom_metadata_map={}".format(meta.custom_metadata_map))
```

**Comment by Mddct** [comment]
onnxruntime support openvino provider。Do we have any plan to support openvino？

**Comment by robin1001** [comment]
> onnxruntime support openvino provider。Do we have any plan to support openvino？

openvino is not in our paln. The user could switch to openvino backend if they do want it.

**Review by robin1001 — APPROVED** [review]

**Comment by lzhin** [comment]
> onnxruntime support openvino provider。Do we have any plan to support openvino？

hello, could you help me look at this [problem](https://github.com/microsoft/onnxruntime/issues/11257)?
